File: backend/services/event_service.py
"""Business rules for listing events."""

from models.event import Event
from repositories.event_repository import EventRepository


class EventService:
    def __init__(self, repository: EventRepository | None = None) -> None:
        self.repository = repository or EventRepository()

    def get_events(self) -> list[dict]:
        events = [Event.from_item(item) for item in self.repository.list_events()]
        return [self._with_calculated_status(event) for event in sorted(events, key=lambda item: (item.date, item.time))]

    @staticmethod
    def _with_calculated_status(event: Event) -> dict:
        data = event.to_api()
        if event.available_seats == 0:
            data["status"] = "SOLD_OUT"
        elif event.available_seats <= max(1, event.total_seats // 10):
            data["status"] = "LIMITED"
        else:
            data["status"] = "AVAILABLE"
        return data

File: backend/services/eventService.js
/**
 * Business rules for listing events.
 */

const { Event } = require("../models/event");
const { EventRepository } = require("../repositories/eventRepository");

class EventService {
  /**
   * @param {Object} [options]
   * @param {Object} [options.repository] - EventRepository override (for testing).
   */
  constructor({ repository } = {}) {
    this.repository = repository || new EventRepository();
  }

  /**
   * Retrieve all events, sorted by date and time, with computed status.
   *
   * @returns {Promise<Object[]>}
   */
  async getEvents() {
    const items = await this.repository.listEvents();
    const events = items.map((item) => Event.fromItem(item));

    events.sort((a, b) => {
      if (a.date !== b.date) return a.date < b.date ? -1 : 1;
      return a.time < b.time ? -1 : a.time > b.time ? 1 : 0;
    });

    return events.map((event) => EventService._withCalculatedStatus(event));
  }

  /**
   * Compute a display status based on remaining seat availability.
   *
   * @param {Event} event
   * @returns {Object} - API-ready event object with computed status.
   */
  static _withCalculatedStatus(event) {
    const data = event.toApi();
    if (event.availableSeats === 0) {
      data.status = "SOLD_OUT";
    } else if (event.availableSeats <= Math.max(1, Math.floor(event.totalSeats / 10))) {
      data.status = "LIMITED";
    } else {
      data.status = "AVAILABLE";
    }
    return data;
  }
}

module.exports = { EventService };
